api/instance.go

```diff
@@ -0,0 +1,117 @@
+			return nil, notFoundError("Instance not found")
+		}
+		return nil, internalServerError("Database error loading instance").WithInternalError(err)
+	}
+	return gcontext.WithInstance(r.Context(), i), nil
+}
+
+func (a *API) GetAppManifest(w http.ResponseWriter, r *http.Request) error {
+	// TODO update to real manifest
+	return sendJSON(w, http.StatusOK, map[string]string{
+		"version":     a.version,
+		"name":        "GoCommerce",
+		"description": "GoCommerce is a flexible Ecommerce API for JAMStack sites",
+	})
+}
+
+type InstanceRequestParams struct {
+	UUID       string              `json:"uuid"`
+	BaseConfig *conf.Configuration `json:"config"`
+}
+
+type InstanceResponse struct {
+	models.Instance
+	Endpoint string `json:"endpoint"`
+	State    string `json:"state"`
+}
+
+func (a *API) CreateInstance(w http.ResponseWriter, r *http.Request) error {
+	params := InstanceRequestParams{}
+	if err := json.NewDecoder(r.Body).Decode(&params); err != nil {
+		return badRequestError("Error decoding params: %v", err)
+	}
+
+	_, err := models.GetInstanceByUUID(a.db, params.UUID)
+	if err != nil {
+		if !models.IsNotFoundError(err) {
+			return internalServerError("Database error looking up instance").WithInternalError(err)
+		}
+	} else {
+		return badRequestError("An instance with that UUID already exists")
+	}
+
+	i := models.Instance{
+		ID:         uuid.NewRandom().String(),
+		UUID:       params.UUID,
+		BaseConfig: params.BaseConfig,
+	}
+	if err = models.CreateInstance(a.db, &i); err != nil {
+		return internalServerError("Database error creating instance").WithInternalError(err)
+	}
+
+	resp := InstanceResponse{
+		Instance: i,
+		Endpoint: a.config.API.Endpoint,
+		State:    "active",
+	}
+	return sendJSON(w, http.StatusCreated, resp)
+}
+
+func (a *API) GetInstance(w http.ResponseWriter, r *http.Request) error {
+	i := gcontext.GetInstance(r.Context())
+	return sendJSON(w, http.StatusOK, i)
+}
+
+func (a *API) UpdateInstance(w http.ResponseWriter, r *http.Request) error {
+	i := gcontext.GetInstance(r.Context())
+
+	params := InstanceRequestParams{}
+	if err := json.NewDecoder(r.Body).Decode(&params); err != nil {
+		return badRequestError("Error decoding params: %v", err)
+	}
+
+	if params.BaseConfig != nil {
+		if err := mergo.MergeWithOverwrite(i.BaseConfig, params.BaseConfig); err != nil {
+			return internalServerError("Error merging instance configurations").WithInternalError(err)
+		}
+	}
+
+	if err := models.UpdateInstance(a.db, i); err != nil {
+		return internalServerError("Database error updating instance").WithInternalError(err)
+	}
+	return sendJSON(w, http.StatusOK, i)
+}
+
+func (a *API) DeleteInstance(w http.ResponseWriter, r *http.Request) error {
+	i := gcontext.GetInstance(r.Context())
+	if err := models.DeleteInstance(a.db, i); err != nil {
+		return internalServerError("Database error deleting instance").WithInternalError(err)
+	}
+
+	// TODO do we delete everything associated with an instance too?
```

Comment:
file a ticket

Reply:
Let's do hard remove for now. If we want to change it to soft delete later, we can do that all at once.

Reply:
https://github.com/netlify/gocommerce/issues/85